scripts/Register-C4bEndpoint.ps1

```diff
@@ -1,12 +1,129 @@
+    [Parameter()]
+    [Switch]
+    $IncludePackageTools,
+    
+    # Allows for the application of user-defined configuration that is applied after the base configuration.
+    # Can override base configuration with this parameter
+    [Parameter()]
+    [Hashtable]
+    $AdditionalConfiguration,
+
+    # Allows for the toggling of additonal features that is applied after the base configuration.
+    # Can override base configuration with this parameter
+    [Parameter()]
+    [Hashtable]
+    $AdditionalFeatures,
+
+    # Allows for the installation of additional packages after the system base packages have been installed.
+    [Parameter()]
+    [Hashtable[]]
+    $AdditionalPackages,
+
+    # Allows for the addition of alternative sources after the base conifguration  has been applied.
+    # Can override base configuration with this parameter
+    [Parameter()]
+    [Hashtable[]]
+    $AdditionalSources,
+
+    [Parameter()]
+    [Switch]
+    $TrustCertificate
+)
+
+# Touch NOTHING below this line
+begin {
+    # Set up our downloader
+    $downloader = [System.Net.WebClient]::new()
+
+    #setup proxy if required
+    if ($ProxyUrl) {
+        $params.add('ProxyUrl', $ProxyUrl)
+        $proxy = [System.Net.WebProxy]::new($ProxyUrl, $true <#bypass on local#>)
+   
+        if ($ProxyCredential) {
+            $params.Add('ProxyCredential', $ProxyCredential)
+            $proxy.Credentials = $ProxyCredential
+        }
+
+        $downloader.Proxy = $proxy
+    }
+
+    $downloader.Credentials = $RepositoryCredential
+
+}
+
+process {
+    #If we use a Self-Signed certificate, we need to explicity trust it
+
+    if ($TrustCertificate) {
+        Invoke-Expression ($downloader.DownloadString("http://$($Fqdn):80/Import-ChocoServerCertificate.ps1"))
+    }
+    #Once we trust the SSL certificate, we can start onboarding
+    $RepositoryUrl = "https://$($fqdn):8443/repository/ChocolateyInternal/index.json"
+
+    #Initialize params hashtable so we can add to it if needed later
+    $params = @{
+        Credential    = $RepositoryCredential
+        ClientSalt    = $ClientCommunicationSalt
+        ServiceSalt   = $ServiceCommunicationSalt
+        RepositoryUrl = $RepositoryUrl
+    }
+
+    switch ($true) {
+        $PSBoundParameters.ContainsKey('AdditionalConfiguration') {
+            $params.Add('AdditionalConfiguration', $AdditionalConfiguration)
+        }
+        $PSBoundParameters.ContainsKey('AdditionalFeatures') {
+            $params.add('AdditionalFeatures', $AdditionalFeatures)
+        }
+
+        $PSBoundParameters.ContainsKey('AdditionalPackages') {
+            $params.Add('AdditionalPackages', $AdditionalPackages)
+        }
+
+        $PSBoundParameters.ContainsKey('AdditionalSources') {
+            $params.Add('AdditionalSources', $AdditionalSources)
+        }
+
+        $PSBoundParameters.ContainsKey('IncludePackageTools') {
+            $params.Add('IncludePackageTools', $IncludePackageTools)
+        }
+    }
```

Comment:
Instead of all this, given we control the params, is it worth literally just parsing and passing PSBoundParams? I think the only thing that would need changing is FQDN/RepositoryUrl.

```suggestion
    $params = $PSBoundParameters

    foreach ($Parameter in @("FQDN", "TrustCertificate")) {
        $null = $params.Remove($Parameter)
    }

    $params += @{
        RepositoryUrl = $RepositoryUrl
    }
```

Please note, I've not tested this (and there's probably more properties to nuke), but it seems like there's less code to maintain with this approach.